File: Medium/90.子集-ii.cpp
/*
 * @lc app=leetcode.cn id=90 lang=cpp
 *
 * [90] 子集 II
 */

// @lc code=start
class Solution
{
public:
    //递归回溯辅助算法
    //注意temp和res都要加上&引用地址符号
    void dfs(vector<int> &temp, vector<vector<int>> &res, int cur, vector<int> nums)
    {
        //每一次回溯递归都要将temp加入到res中
        res.push_back(temp);
        for (int i = cur; i < nums.size(); i++)
        {
            //选择了当前的数字
            temp.push_back(nums[i]);
            //递归回溯
            dfs(temp, res, i + 1, nums);
            //不选择当前数字
            temp.pop_back();
            //去掉重复数字
            while (i < nums.size() - 1 && nums[i] == nums[i + 1])
            {
                i++;
            }
        }
    }

    vector<vector<int>> subsetsWithDup(vector<int> &nums)
    {
        vector<int> temp;
        vector<vector<int>> res;
        //先排序
        sort(nums.begin(), nums.end());
        dfs(temp, res, 0, nums);
        return res;
    }
};
// @lc code=end

File: Medium/90.子集-ii.go
/*
 * @lc app=leetcode.cn id=90 lang=golang
 *
 * [90] 子集 II
 */

// @lc code=start
var res [][]int

func subsetsWithDup(nums []int) [][]int {
	res = make([][]int, 0)
	// 排序可以保证重复数字在一起
	sort.Ints(nums)
	dfs([]int{}, nums, 0)
	return res
}

//
func dfs(temp, nums []int, start int) {
	tmp := make([]int, len(temp))
	copy(tmp, temp)
	res = append(res, tmp)
	for i := start; i < len(nums); i++ {
		// 若发现没有选择上一个数，且当前数字与上一个数相同，则可以跳过当前生成的子集。
		if i > start && nums[i] == nums[i-1] {
			continue
		}
		// 数字加入
		temp = append(temp, nums[i])
		// 递归dfs
		dfs(temp, nums, i+1)
		// 若发现没有选择上一个数，且当前数字与上一个数相同，则可以跳过当前生成的子集。
		temp = temp[:len(temp)-1]
	}
}

// @lc code=end

